lib/Controller/MessageApiController.php

```diff
@@ -0,0 +1,93 @@
+<?php
+
+declare(strict_types=1);
+/**
+ * SPDX-FileCopyrightText: 2024 Nextcloud GmbH and Nextcloud contributors
+ * SPDX-License-Identifier: AGPL-3.0-only
+ */
+namespace OCA\Mail\Controller;
+
+use OCA\Mail\Exception\ClientException;
+use OCA\Mail\Exception\ServiceException;
+use OCA\Mail\Http\AttachmentDownloadResponse;
+use OCA\Mail\IMAP\IMAPClientFactory;
+use OCA\Mail\Service\AccountService;
+use OCA\Mail\Service\AliasesService;
+use OCA\Mail\Service\Attachment\AttachmentService;
+use OCA\Mail\Service\MailManager;
+use OCA\Mail\Service\OutboxService;
+use OCP\AppFramework\Db\DoesNotExistException;
+use OCP\AppFramework\Http;
+use OCP\AppFramework\Http\Attribute\BruteForceProtection;
+use OCP\AppFramework\Http\Attribute\NoAdminRequired;
+use OCP\AppFramework\Http\Attribute\NoCSRFRequired;
+use OCP\AppFramework\Http\DataResponse;
+use OCP\AppFramework\OCSController;
+use OCP\AppFramework\Utility\ITimeFactory;
+use OCP\IRequest;
+use OCP\IUserManager;
+use Psr\Log\LoggerInterface;
+
+class MessageApiController extends OCSController {
+
+	private ?string $userId;
+
+	public function __construct(
+		string $appName,
+		$UserId,
+		IRequest $request,
+		private IUserManager $userManager,
+		private AccountService $accountService,
+		private AliasesService $aliasesService,
+		private AttachmentService $attachmentService,
+		private OutboxService $outboxService,
+		private MailManager $mailManager,
+		private IMAPClientFactory $clientFactory,
+		private LoggerInterface $logger,
+		private ITimeFactory $time,
+	) {
+		parent::__construct($appName, $request);
+		$this->userId = $UserId;
+	}
+
+	/**
+	 * @param int $id
+	 * @return DataResponse
+	 */
+	#[BruteForceProtection('mailGetMessage')]
+	#[NoAdminRequired]
+	#[NoCSRFRequired]
+	public function get(int $id): DataResponse {
+		try {
+			$message = $this->mailManager->getMessage($this->userId, $id);
+			$mailbox = $this->mailManager->getMailbox($this->userId, $message->getMailboxId());
+			$account = $this->accountService->find($this->userId, $mailbox->getAccountId());
+		} catch (ClientException | DoesNotExistException $e) {
+			$this->logger->error('Message, Account or Mailbox not found', ['exception' => $e->getMessage()]);
+			return new DataResponse('Forbidden', Http::STATUS_FORBIDDEN);
+		}
+
+		$client = $this->clientFactory->getClient($account);
+		try {
+			$source = $this->mailManager->getSource(
+				$client,
+				$account,
+				$mailbox->getName(),
+				$message->getUid()
+			);
+		} catch (ServiceException $e) {
+			$this->logger->error('Message not found on IMAP or mail server went away', ['exception' => $e->getMessage()]);
+			return new DataResponse('Not found', Http::STATUS_NOT_FOUND);
+		} finally {
+			$client->logout();
+		}
+
+		$messageContent = new AttachmentDownloadResponse(
+			$source,
+			$message->getSubject() . '.eml',
+			'message/rfc822',
+		);
```

Comment:
I would recommend returning the message as a Json object instead of a raw email string, this way the consuming service does not have to re-phrase the message in again to use/read its contents, like we do in the UI.

Something like this... (pulled from a UI request)

```
{
    "uid": 254,
    "messageId": "\u003C3f488a7111b4a63cf073ec0ec4f86ff9@ksamail.ca\u003E",
    "from": [
        {
            "label": "User One via Nextcloud",
            "email": "ncdev@ksamail.ca"
        }
    ],
    "to": [
        {
            "label": "User Two",
            "email": "user2@ksamail.ca"
        }
    ],
    "replyTo": [
        {
            "label": "User One",
            "email": "user1@ksamail.ca"
        }
    ],
    "cc": [],
    "bcc": [],
    "subject": "Invitation: Recurrance Test",
    "dateInt": 1718673419,
    "flags": {
        "seen": true,
        "flagged": false,
        "answered": false,
        "deleted": false,
        "draft": false,
        "forwarded": false,
        "hasAttachments": false,
        "mdnsent": false,
        "important": false
    },
    "hasHtmlBody": true,
    "dispositionNotificationTo": "",
    "hasDkimSignature": false,
    "unsubscribeUrl": null,
    "isOneClickUnsubscribe": false,
    "unsubscribeMailto": null,
    "body": "==== Email Body =====",
    "attachments": [
        {
            "id": "2",
            "messageId": 254,
            "fileName": "event.ics",
            "mime": "text\/calendar",
            "size": 1566,
            "cid": null,
            "disposition": "inline",
            "downloadUrl": "https:\/\/localhost\/apps\/mail\/api\/messages\/876\/attachment\/2",
            "mimeUrl": "\/apps\/theming\/img\/core\/filetypes\/text-calendar.svg?v=704c4934",
            "isImage": false,
            "isCalendarEvent": true
        }
    ],
    "accountId": 2,
    "mailboxId": 7,
    "databaseId": 876,
    "isSenderTrusted": false,
    "smime": {
        "isSigned": false,
        "signatureIsValid": null,
        "isEncrypted": false
    }
}
```

Reply:
Thanks for the suggestion - I think a combo of both parsed data and the raw email might be beneficial so that the workflow engine can proceed with whatever data they think is best. Good tip!

Reply:
You're welcome.

You could add a raw flag to the get request, that way the consuming server can choose...

```
public function get(int $id, bool $raw = false) {

}
```

Or  something like this....

```
public function get(int $id, string $format = 'json') {

}
```

Reply:
I think it might be better to split the get into two separate endpoints even